resolve manually set basetemp to a absolute path

Hi Guys,

We're experiencing some issues with new tmpdir implementation using pathlib. Starting from pytest 3.9.0 our tests started to fail. 

We are using test class and a pre_process fixture, which prepares some data for the other tests within the class. In the test methods we're using the pre_process fixture (to access the data) as well as the tmpdir pytest builtin fixture to store some outcomes of the tests. Minimal failing example is pasted below:
```
import pytest, os

class TestDummy():
    @pytest.fixture(scope="class")
    def pre_process(self, tmpdir_factory):
        path_with_data = tmpdir_factory.mktemp('data')
        return path_with_data

    @pytest.mark.parametrize("param", range(2))
    def test_dummy(self, tmpdir, pre_process, param):
        os.chdir(pre_process)
        # further processing

    def test_dummy2(self, tmpdir, pre_process):
        os.chdir(pre_process)
        # further processing

```
Because of the directory change in test_dummy, next invocation of the tests fail due to pathlib error, as below:
```____________________________________________________________________________________________________ ERROR at setup of TestDummy.test_dummy[1] _____________________________________________________________________________________________________

self = PosixPath('tmp')

    def iterdir(self):
        """Iterate over the files in this directory.  Does not yield any
        result for the special paths '.' and '..'.
        """
        if self._closed:
            self._raise_closed()
>       for name in self._accessor.listdir(self):

../../../.pyenv/versions/3.6.5/lib/python3.6/pathlib.py:1079: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pathobj = PosixPath('tmp'), args = ()

    @functools.wraps(strfunc)
    def wrapped(pathobj, *args):
>       return strfunc(str(pathobj), *args)
E       FileNotFoundError: [Errno 2] No such file or directory: 'tmp'

../../../.pyenv/versions/3.6.5/lib/python3.6/pathlib.py:387: FileNotFoundError
```

and in fact we end up with 1 pass and 2 errors in this particular case.

We currently have workaround for this issue (store the cwd before changing it and restore it before the test end), but I'd like to ask you if this is how it should behave? I'd expect that state of one test does not influence the other test's state. 

The issue is observable if no class hierarchy is used as well. 

Another workaround (a little bit nasty) is to reset internal pytest variable in pre_process macro:
```@pytest.fixture(scope="module")
def pre_process(tmpdir_factory):
    path_with_data = tmpdir_factory.mktemp('data')
    tmpdir_factory._tmppath_factory._basetemp = None
    return path_with_data```

Environment details:
Python 3.6.5 (via pyenv), 4.15.0-29-generic #31-Ubuntu SMP Tue Jul 17 15:37:27 UTC 2018 i686 i686 i686 GNU/Linux
```
$ pip list
Package        Version
-------------- -------
atomicwrites   1.2.1  
attrs          18.2.0 
more-itertools 4.3.0  
pip            18.1   
pluggy         0.8.0  
py             1.7.0  
pytest         4.0.0  
setuptools     40.6.2 
six            1.11.0 ```

- [X] Include a detailed description of the bug or suggestion
- [X] `pip list` of the virtual environment you are using:
- [X] pytest and operating system versions:
- [X] Minimal example if possible - already in the text

Sure:
```
$ pytest test.py 
=============================================================================================================== test session starts ================================================================================================================
platform linux -- Python 3.6.5, pytest-4.0.0, py-1.7.0, pluggy-0.8.0
rootdir: /home/mrusi, inifile: pytest.ini
collected 3 items                                                                                                                                                                                                                                  

test.py .EE                                                                                                                                                                                                                                  [100%]

====================================================================================================================== ERRORS ======================================================================================================================
____________________________________________________________________________________________________ ERROR at setup of TestDummy.test_dummy[1] _____________________________________________________________________________________________________

self = PosixPath('tmp')

    def iterdir(self):
        """Iterate over the files in this directory.  Does not yield any
        result for the special paths '.' and '..'.
        """
        if self._closed:
            self._raise_closed()
>       for name in self._accessor.listdir(self):

../../../.pyenv/versions/3.6.5/lib/python3.6/pathlib.py:1079: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pathobj = PosixPath('tmp'), args = ()

    @functools.wraps(strfunc)
    def wrapped(pathobj, *args):
>       return strfunc(str(pathobj), *args)
E       FileNotFoundError: [Errno 2] No such file or directory: 'tmp'

../../../.pyenv/versions/3.6.5/lib/python3.6/pathlib.py:387: FileNotFoundError
_____________________________________________________________________________________________________ ERROR at setup of TestDummy.test_dummy2 ______________________________________________________________________________________________________

self = PosixPath('tmp')

    def iterdir(self):
        """Iterate over the files in this directory.  Does not yield any
        result for the special paths '.' and '..'.
        """
        if self._closed:
            self._raise_closed()
>       for name in self._accessor.listdir(self):

../../../.pyenv/versions/3.6.5/lib/python3.6/pathlib.py:1079: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pathobj = PosixPath('tmp'), args = ()

    @functools.wraps(strfunc)
    def wrapped(pathobj, *args):
>       return strfunc(str(pathobj), *args)
E       FileNotFoundError: [Errno 2] No such file or directory: 'tmp'

../../../.pyenv/versions/3.6.5/lib/python3.6/pathlib.py:387: FileNotFoundError
================================================================================================================= warnings summary =================================================================================================================
/home/mrusi/.pyenv/versions/3.6.5/lib/python3.6/site-packages/_pytest/resultlog.py:39
  /home/mrusi/.pyenv/versions/3.6.5/lib/python3.6/site-packages/_pytest/resultlog.py:39: RemovedInPytest4Warning: --result-log is deprecated and scheduled for removal in pytest 4.0.
  See https://docs.pytest.org/en/latest/usage.html#creating-resultlog-format-files for more information.
    _issue_config_warning(RESULT_LOG, config)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================================================================================== 1 passed, 1 warnings, 2 error in 0.09 seconds ===================================================================================================

```

--- a/changelog/4425.bugfix.rst
+++ b/changelog/4425.bugfix.rst
@@ -1,0 +1,1 @@
+Ensure we resolve the absolute path when the given ``--basetemp`` is a relative path.

--- a/src/_pytest/monkeypatch.py
+++ b/src/_pytest/monkeypatch.py
@@ -13,8 +13,9 @@ import six
 
 import pytest
 from _pytest.fixtures import fixture
+from _pytest.pathlib import Path
 
-RE_IMPORT_ERROR_NAME = re.compile("^No module named (.*)$")
+RE_IMPORT_ERROR_NAME = re.compile(r"^No module named (.*)$")
 
 
 @fixture
@@ -267,6 +268,9 @@ class MonkeyPatch(object):
             self._cwd = os.getcwd()
         if hasattr(path, "chdir"):
             path.chdir()
+        elif isinstance(path, Path):
+            # modern python uses the fspath protocol here LEGACY
+            os.chdir(str(path))
         else:
             os.chdir(path)
 

--- a/src/_pytest/tmpdir.py
+++ b/src/_pytest/tmpdir.py
@@ -26,7 +26,9 @@ class TempPathFactory(object):
 
     The base directory can be configured using the ``--basetemp`` option."""
 
-    _given_basetemp = attr.ib()
+    _given_basetemp = attr.ib(
+        convert=attr.converters.optional(lambda p: Path(p).resolve())
+    )
     _trace = attr.ib()
     _basetemp = attr.ib(default=None)
 
@@ -53,7 +55,7 @@ class TempPathFactory(object):
         """ return base temporary directory. """
         if self._basetemp is None:
             if self._given_basetemp is not None:
-                basetemp = Path(self._given_basetemp)
+                basetemp = self._given_basetemp
                 ensure_reset_dir(basetemp)
             else:
                 from_env = os.environ.get("PYTEST_DEBUG_TEMPROOT")

--- a/src/_pytest/tmpdir.py
+++ b/src/_pytest/tmpdir.py
@@ -10,6 +10,7 @@ import warnings
 
 import attr
 import py
+import six
 
 import pytest
 from .pathlib import ensure_reset_dir
@@ -28,8 +29,11 @@ class TempPathFactory(object):
 
     _given_basetemp = attr.ib(
         # using os.path.abspath() to get absolute path instead of resolve() as it
-        # does not work the same in all platforms
-        convert=attr.converters.optional(lambda p: Path(os.path.abspath(p)))
+        # does not work the same in all platforms; there's Path.absolute(), but it is not
+        # public (see https://bugs.python.org/issue25012)
+        convert=attr.converters.optional(
+            lambda p: Path(os.path.abspath(six.text_type(p)))
+        )
     )
     _trace = attr.ib()
     _basetemp = attr.ib(default=None)

--- a/src/_pytest/tmpdir.py
+++ b/src/_pytest/tmpdir.py
@@ -27,7 +27,9 @@ class TempPathFactory(object):
     The base directory can be configured using the ``--basetemp`` option."""
 
     _given_basetemp = attr.ib(
-        convert=attr.converters.optional(lambda p: Path(p).resolve())
+        # using os.path.abspath() to get absolute path instead of resolve() as it
+        # does not work the same in all platforms
+        convert=attr.converters.optional(lambda p: Path(os.path.abspath(p)))
     )
     _trace = attr.ib()
     _basetemp = attr.ib(default=None)

--- a/src/_pytest/tmpdir.py
+++ b/src/_pytest/tmpdir.py
@@ -29,8 +29,8 @@ class TempPathFactory(object):
 
     _given_basetemp = attr.ib(
         # using os.path.abspath() to get absolute path instead of resolve() as it
-        # does not work the same in all platforms; there's Path.absolute(), but it is not
-        # public (see https://bugs.python.org/issue25012)
+        # does not work the same in all platforms (see #4427)
+        # Path.absolute() exists, but it is not public (see https://bugs.python.org/issue25012)
         convert=attr.converters.optional(
             lambda p: Path(os.path.abspath(six.text_type(p)))
         )

--- a/testing/test_tmpdir.py
+++ b/testing/test_tmpdir.py
@@ -4,6 +4,7 @@ from __future__ import print_function
 
 import sys
 
+import attr
 import six
 
 import pytest
@@ -25,12 +26,29 @@ def test_ensuretemp(recwarn):
     assert d1.check(dir=1)
 
 
+@attr.s
+class FakeConfig(object):
+    basetemp = attr.ib()
+    trace = attr.ib(default=None)
+
+    @property
+    def trace(self):
+        return self
+
+    def get(self, key):
+        return lambda *k: None
+
+    @property
+    def option(self):
+        return self
+
+
 class TestTempdirHandler(object):
-    def test_mktemp(self, testdir):
+    def test_mktemp(self, tmp_path):
+
         from _pytest.tmpdir import TempdirFactory, TempPathFactory
 
-        config = testdir.parseconfig()
-        config.option.basetemp = testdir.mkdir("hello")
+        config = FakeConfig(tmp_path)
         t = TempdirFactory(TempPathFactory.from_config(config))
         tmp = t.mktemp("world")
         assert tmp.relto(t.getbasetemp()) == "world0"
@@ -39,6 +57,15 @@ class TestTempdirHandler(object):
         tmp2 = t.mktemp("this")
         assert tmp2.relto(t.getbasetemp()).startswith("this")
         assert tmp2 != tmp
+
+    @pytest.mark.issue(4425)
+    def test_tmppath_relative_basetemp_absolute(self, tmp_path, monkeypatch):
+        from _pytest.tmpdir import TempPathFactory
+
+        monkeypatch.chdir(tmp_path)
+        config = FakeConfig("hello")
+        t = TempPathFactory.from_config(config)
+        assert t.getbasetemp() == (tmp_path / "hello")
 
 
 class TestConfigTmpdir(object):

--- a/testing/test_tmpdir.py
+++ b/testing/test_tmpdir.py
@@ -65,7 +65,7 @@ class TestTempdirHandler(object):
         monkeypatch.chdir(tmp_path)
         config = FakeConfig("hello")
         t = TempPathFactory.from_config(config)
-        assert t.getbasetemp() == (tmp_path / "hello")
+        assert t.getbasetemp().resolve() == (tmp_path / "hello").resolve()
 
 
 class TestConfigTmpdir(object):

